Revisión de Pkg building.es.rmd es (#581)
ropensci/dev_guide: pkg_building.es.Rmd

# Guía de empaquetado {#building}

```{block, type='summaryblock'}
rOpenSci acepta paquetes que cumplen nuestras recomendaciones a través de un proceso fluido de [Revisión por pares de software] (#whatissoftwarereview). Para garantizar un estilo coherente en todas nuestras herramientas, hemos escrito este capítulo en el que se destacan recomendaciones para el desarrollo de paquetes. Por favor, también lee y aplica nuestro [capítulo sobre integración continua (CI)](#ci). En la tercera sección de este libro, que comienza con [un capítulo sobre la colaboración](#colaboración), se ofrece una guía para continuar luego del proceso de revisión.

> Recomendamos fuertemente a quienes desarrollen paquetes que lean el libro de Hadley Wickham y Jenny Bryan sobre el desarrollo de paquetes. Es conciso pero exhaustivo y está disponible [gratis en línea](https://r-pkgs.org/). Algunas partes de nuetra guía son redundantes con respecto a otros recursos, pero destaca las recomendaciones de rOpenSci.

Reviewer: Estuve buscando sinónimos perotodos suenan peor que fuertemente.

Para leer por qué vale la pena enviar un paquete a rOpenSci siguiendo las recomendaciones, echa un vistazo a [razones para enviar un paquete](#whysubmit). 

```

## Nombre del paquete y metadatos

### Cómo elegir el nombre de tu paquete

- Recomendamos fuertemente nombres cortos y descriptivos en minúsculas. Si tu paquete tiene que ver con uno o más servicios comerciales, asegúrate de que el nombre no infringe las recomendaciones de la marca. Puedes comprobar si el nombre de tu paquete está disponible, es informativo y no es ofensivo utilizando el [paquete `available`](https://github.com/ropenscilabs/available). En particular, *no* elijas un nombre de paquete que ya se utilice en CRAN o en Bioconductor.

> - Un nombre de paquete menos común podría ser más fácil de seguir (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).

Reviewer: Creo que en vez de "seguir", "rastrear" es más claro. Interpreto que un nombre único-poco común hace fácil rastrear/buscar donde se usa el paquete. 

Sobre lo segundo, la primera vez me pareció razonable. Ahora no le encuentro sentido, ni siquiera a la versión en inglés: Obviously a too unique package name might make the package less discoverable (e.g. it might be an argument for naming your package [geojson](https://github.com/ropensci/geojson)).

> - Un nombre de paquete menos común podría ser más fácil de seguir (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).

Reviewer 2: Tal vez el ejemplo no tiene tanto sentido para nosotras.  Entiendo que está hablando de justificar porqué seleccionaste un nombre taaaan general. Algo como, le pongo geojson porque todo el mundo va a buscarlo asi. Tal vez es necesario hacer eso más explicito en el texto en ambos idiomas y no dejarlo a que el lector sepa que geojson es un nombre no común o muy general (si mi interpretación es correcta)

> - Un nombre de paquete menos común podría ser más fácil de seguir (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).

Reviewer: Me gusta tu sugerencia. Solo me preocupa cambiar la estructura del párrafo (oraciones, paréntesis) porque no se como se lleva eso con el futuro sistema que actualice las traducciones. Esta es la razón principal por la cuál hemos sido bastante conservadores hasta ahora.

> - Un nombre de paquete menos común podría ser más fácil de seguir (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).

Reviewer 2: Tenés razón.  Interesante problema. Entonces algo como  _Un nombre de paquete menos común podría ser más fácil de encontrar (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas)_

> - Un nombre de paquete menos común podría ser más fácil de seguir (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).

Author: ```suggestion
- Un nombre de paquete menos común podría ser más fácil de encontrar (para que tú y en rOpenSci evaluemos el uso del paquete) y de buscar (para que quienes lo usen lo encuentren y busquen en Google sus preguntas). Evidentemente, un nombre *demasiado* único puede hacer que el paquete sea más difícil de encontrar (por ejemplo, podría ser un argumento para nombrar tu paquete [geojson](https://github.com/ropensci/geojson)).
```

## API del paquete

### Nombres de funciones y argumentos

- Los nombres de las funciones y los argumentos deben elegirse para que funcionen juntos y formen una API de programación común y lógica que sea fácil de leer y autocompletar.
  
  - Considera un esquema de nomenclatura `objecto_verbo()` para las funciones de tu paquete que toman un tipo de datos común o interactúan con una API común. `objecto` se refiere a los datos/API y `verbo` a la acción principal.  Este esquema ayuda a evitar conflictos de nombres con paquetes que pueden tener verbos similares, y hace que el código sea legible y fácil de autocompletar.  Por ejemplo, en **stringi** las funciones que empiezan con `stri_` manipulan cadenas de texto (`stri_join()`, `stri_sort()`y en **googlesheets** las funciones que empiezan con `gs_` son llamadas a la API de Google Sheets (`gs_auth()`, `gs_user()`, `gs_download()`).

- Para las funciones que manipulan un objeto o dato y devuelven un objeto/dato del mismo tipo, haz que el objeto o dato sea el primer argumento de la función para mejorar la compatibilidad con el operador pipe (`%>%` o `|>`).

> - Recomendamos fuertemente usar `snake_case` por sobre todos los otros estilos, a menos que estés usando funciones de un paquete que ya se utilice ampliamente.

Author: Sí, sería para discutir en otro lado. 

Me parece que snake_case es mejor para accesibilidad porque no depende de la capitalización, que el lector de pantalla no lee. Entonces una persona ciega escucha "CamelCase" y no sabe si es "camel case", "Camel Case", "camelCase" u otra cosa. Mientras tanto, usar snake_case siempre en minúscula es más claro porque el lector lee "snake underscore case".

  - Cualquier configuración adicional que se requiera (tokens de autenticación, etc).
  - Breve demostración de uso.
  - Si aplica, cómo se compara el paquete con otros paquetes similares y/o cómo se relaciona con otros paquetes.
  - Información sobre cómo citar, es decir, muestra a las personas el formato de cita preferida en el *README* añadiendo el texto "así es como se puede citar mi paquete". Véase, por ejemplo [el *README* de ecmwfr](https://github.com/bluegreen-labs/ecmwfr#how-to-cite-this-package-in-your-article).

Si utilizas otra etiqueta de estado del repo, como el [ciclo de vida](https://www.tidyverse.org/lifecycle/) por favor, añade también una etiqueta de [repostatus.org](https://www.repostatus.org/) . [Ejemplo de un *README* de un repo con dos insignias de estado](https://github.com/ropensci/ijtiff#ijtiff-).

- Una vez que hayas enviado el paquete y haya completado la revisión editorial, añade la insignia de revisión por pares como

```
[![](https://badges.ropensci.org/<issue_id>_status.svg)](https://github.com/ropensci/software-review/issues/<issue_id>)
```

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Reviewer: Para mi es masculino. @eliocamp?

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Author: Me da lo mismo. Creo que uso ambos géneros medio aleatoriamente, que es la peor opción. 

Google Ngram no encuentra [ninguna referencia](https://books.google.com/ngrams/graph?content=un+issue%2Cuna+issue%2Cel+issue%2Cla+issue&year_start=1800&year_end=2019&corpus=21&smoothing=0) a "la issue".

Armé esta encuesta en mastodon: https://mastodon.social/@eliocamp/109638310946336510

¿Se puede armar una similar en twitter?

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Reviewer 2: Estoy viendo si puedo crearla en el tweeter de rOpenSci

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Reviewer 2: No se como crear encuentas usando tweet deck.  Pero tu encuesta consiguió 27 votos y masculino lleva una marcada delantera.  Asi que lo ponemos en masculino y lo agregamos a la guia.  Chas gracias!!

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Author: Sí, de acuerdo.

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Author: Veo que el PR ya está mergeado pero no me queda claro si esta parte quedó con "la issue" o cambió a "el issue".

> donde issue\_id es el número de la issue en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de issue de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (issue etiquetada como "aprobado" y cerrado), y estará vinculado con la issue de la revisión.

Reviewer 2: ```suggestion
donde issue\_id es el número del _issue _en el repositorio donde se hizo la revisión del software. Por ejemplo, la etiqueta de revisión de [`rtimicropem`](https://github.com/ropensci/rtimicropem) utiliza el número 126, ya que es el [número de _issue_ de revisión](https://github.com/ropensci/software-review/issues/126). La etiqueta indicará primero "en revisión" y luego "revisado" una vez que tu paquete haya sido incorporado (_issue_ etiquetado como "aprobado" y cerrado), y estará vinculado con el _issue_ de la revisión.
```

Si tu paquete no tiene ningún logotipo, el [generador de documentación de rOpenSci](#docsropensci) utilizará el logo de rOpenSci en su lugar.

## Autoría {#authorship}

El archivo `DESCRIPTION` de un paquete debe enumerar las personas que participaron de la autoría y quienes colaboraron en el paquete, utilizando la sintaxis `Authors@R` para indicar sus funciones (*author*/*creator*/*contributor*, etc.) si hay más de una persona listada. Utiliza el campo de comentarios para indicar el ID ORCID de cada persona si tuviera (ver [este post](https://ropensci.org/technotes/2018/10/08/orcid/)). Revisa [esta sección de "Writing R Extensions"](https://cran.rstudio.com/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) para más detalles. Si crees que quienes revisaron tu paquete han hecho una contribución sustancial al desarrollo de tu paquete, puedes agregarlas en el campo `Authors@R` usando el tipo de contribución `"rev"`, por ejemplo:

```
    person("Bea", "Hernández", role = "rev",
    comment = "Bea revisó el paquete (v. X.X.XX) para rOpenSci, ver <https://github.com/ropensci/software-review/issues/116>"),
```

> No incluyas a nadie en tu paquete sin antes pedir su consentimiento. Lee más en [esta entrada del blog "Thanking Your Reviewers: Gratitude through Semantic Metadata"](https://ropensci.org/blog/2018/03/16/thanking-reviewers-in-metadata/). Por favor, no agregues al equipo editorial. ¡Su participación y contribución a rOpenSci es suficiente agradecimiento!

Reviewer 2: ```suggestion
No incluyas a nadie en tu paquete sin antes pedir su consentimiento. Lee más en [esta entrada del blog "Thanking Your Reviewers: Gratitude through Semantic Metadata" (Agradeciendo a tus revisores: La gratitud a través de los metadatos semánticos)](https://ropensci.org/blog/2018/03/16/thanking-reviewers-in-metadata/). Por favor, no agregues al equipo editorial. ¡Su participación y contribución a rOpenSci es suficiente agradecimiento!
```

Author: ```suggestion
No incluyas a nadie en tu paquete sin antes pedir su consentimiento. Lee más en [esta entrada del blog *Thanking Your Reviewers: Gratitude through Semantic Metadata* (Agradeciendo las revisiones: La gratitud a través de los metadatos semánticos)](https://ropensci.org/blog/2018/03/16/thanking-reviewers-in-metadata/). Por favor, no agregues al equipo editorial. ¡Su participación y contribución a rOpenSci es suficiente agradecimiento!
```

### Autoría del código incluido en el paquete {#authorship-included-code}

Muchos paquetes incluyen código de otro software. Tanto si se incluyen archivos enteros como funciones individuales de otros paquetes, los paquetes de rOpenSci deben seguir [la *Política de Repositorio* de CRAN](https://cran.r-project.org/web/packages/policies.html):

> La propiedad de los derechos de autor y de propiedad intelectual de todos los componentes del paquete debe ser clara e inequívoca (incluso en la especificación de  autoría en el archivo DESCRIPTION). Cuando el código se copie (o se derive) del trabajo de otras personas (incluso del propio R), hay que tener cuidado de que se conserven las declaraciones de derechos de autor/licencia y no se tergiverse la autoría.
> 

> > Preferiblemente, se utilizaría un campo "Autores@R" con funciones "ctb" para quienes tiene la autoría de dicho código. Como alternativa, el campo "Autor" puede incluir a estas personas como en el rol de colaboración.

Reviewer 2: ```suggestion
> Preferiblemente, se utilizaría un campo "Autores@R" con funciones "ctb" para quienes tiene la autoría de dicho código. Como alternativa, el campo "Autor" puede incluir a estas personas en el rol de colaboración.
```

> 
> Deben respetarse las marcas comerciales.

## Licencia

El paquete debe tener una licencia aceptada por [CRAN](https://svn.r-project.org/R/trunk/share/licenses/license.db) u [OSI](https://opensource.org/licenses).
Para más detalles sobre las licencias, consulta el libro [R packages](https://r-pkgs.org/license.html).

## Testeo

- Todos los paquetes deben pasar `R CMD check`/`devtools::check()` en las principales plataformas.

- Todos los paquetes deben tener un conjunto de *tests* que cubran la funcionalidad principal del paquete. Los *tests* deben cubrir también el comportamiento del paquete en caso de error.

- Es una buena práctica escribir *tests* unitarios para todas las funciones, y para todo el código del paquete en general, asegurando que se cubra las funcionalidades clave. Si la cobertura de los *tests* en tu paquete está por debajo del 75% probablemente requerirá *tests* adicionales o una explicación de la baja cobertura antes de ser enviado para su revisión.

> - Recomendamos utilizar [testthat](https://testthat.r-lib.org/) para escribir los *tests* Intenta escribir *tests* a medida que escribes cada nueva función. Esto responde a la necesidad obvia de tener *tests* adecuadas para el paquete, pero te permite pensar en varias formas en las que una función puede fallar, y *defensivamente* programar en contra esas fallas. [Más información](https://r-pkgs.org/tests.html).

Reviewer 2: ```suggestion
- Recomendamos utilizar [testthat](https://testthat.r-lib.org/) para escribir los *tests*. Intenta escribir *tests* a medida que escribes cada nueva función. Esto responde a la necesidad obvia de tener *tests* adecuados para el paquete, pero te permite pensar en varias formas en las que una función puede fallar, y *defensivamente* programar en contra esas fallas. [Más información](https://r-pkgs.org/tests.html).
```

> - Los *tests* deben ser fáciles de entender. Te sugerimos que leas la entrada del blog [*Why Good Developers Write Bad Unit Tests* (Por qué las personas que desarrollan escriben malos tests)](https://mtlynch.io/good-developers-bad-tests/) de Michael Lynch.

Reviewer 2: ```suggestion
- Los *tests* deben ser fáciles de entender. Te sugerimos que leas el artículo de blog [*Why Good Developers Write Bad Unit Tests* (Por qué las personas que son buenas desarrollando escriben malos tests)](https://mtlynch.io/good-developers-bad-tests/) de Michael Lynch.
```

- Una vez que hayas configurado [la integración continua (CI)](#ci), utiliza el informe de cobertura de tu paquete (revisa [esta sección de nuestro libro](#coverage)) para identificar las líneas no testeadas y añadir más *tests*.

- Aunque utilices la [integración continua](#ci) te recomendamos que ejecutes los *tests* localmente antes de enviar tu paquete ya que a menudo se omiten algunos *tests* (puede que tengas que establecer `Sys.setenv(NOT_CRAN="true")` para asegurarte de que se ejecutan todos los *tests*). Además, te recomendamos que antes de enviar tu paquete utilices el paquete de MangoTheCat [`goodpractice`](https://github.com/MangoTheCat/goodpractice/) para testear tu paquete en busca de posibles fuentes de errores, y ejecutar `spelling::spell_check_package()` para encontrar errores ortográficos en la documentación.

## Ejemplos

- Incluye ejemplos extensos en la documentación. Además de demostrar cómo se utiliza el paquete, pueden ser una forma fácil de testear la funcionalidad del paquete antes de que haya *tests* adecuados. Sin embargo, ten en cuenta que exigimos *tests* en los paquetes contribuidos.

> - Puedes ejecutar los ejemplos con `devtools::run_examples()`. Ten en cuenta que los ejemplos que no estén rodeados de `\dontrun{}` o `\donttest{}` serán ejecutados cuando cuando ejecutes R CMD CHECK o su equivalente (por ejemplo, `devtools::check()`). Consulta la [tabla de resumen](https://roxygen2.r-lib.org/articles/rd.html#functions) en la documentación de roxygen2.

Reviewer 2: ```suggestion
- Puedes ejecutar los ejemplos con `devtools::run_examples()`. Ten en cuenta que los ejemplos que no estén rodeados de `\dontrun{}` o `\donttest{}` serán ejecutados cuando ejecutes R CMD CHECK o su equivalente (por ejemplo, `devtools::check()`). Consulta la [tabla de resumen](https://roxygen2.r-lib.org/articles/rd.html#functions) en la documentación de roxygen2.
```

      ```
      
      (Ten en cuenta que este enfoque [no está universalmente aceptado](https://twitter.com/krlmlr/status/1067856118385381377).)
  
  - Asegúrate de que utilizas la función del paquete donde está definida originalmente y no de un paquete que la re-exporta. Por ejemplo, muchas funciones de **devtools** pueden encontrarse en paquetes especializados más pequeños, como **sessioninfo**. La función `%>%` debe importarse de **magrittr** donde está definida, en lugar de **dplyr** que la reexporta y es mucho más pesado.
  
  - Algunas dependencias proporcionan nombres de funciones y sintaxis más fáciles de interpretar que los nombres de las funciones y la sintaxis de R base. Si ésta es la razón principal para usar una función en una dependencia pesada, considera incluir el código de R base en una función interna bien nombrada en tu paquete. Consulta, por ejemplo, el [script de R de rlang que proporciona funciones con una sintaxis similar a las funciones de purrr](https://github.com/r-lib/rlang/blob/master/R/compat-purrr.R).
  
  - Si las dependencias tienen funcionalidades que se solapan, comprueba si puedes depender sólo de una.
  

>   - Puedes encontrar más consejos sobre la gestión de dependencias [en este post de Scott Chamberlain](https://recology.info/2018/10/limiting-dependencies/).

Reviewer 2: ```suggestion
  - Puedes encontrar más consejos sobre la gestión de dependencias [en este artículo de Scott Chamberlain](https://recology.info/2018/10/limiting-dependencies/).
```

## Estructura de base recomendada

> - Para las consultas HTTP recomendamos utilizar [curl][https://github.com/jeroen/curl/], [crul][https://github.com/ropensci/crul/], [httr][https://httr.r-lib.org] o [httr2][https://httr2.r-lib.org] en vez de [RCurl][https://cran.rstudio.com/web/packages/RCurl/]. El paquete [curl][https://github.com/jeroen/curl/] es mejor es mejor si prefieres los clientes de bajo nivel de HTTP, mientras que [crul][https://github.com/ropensci/crul/] o [httr][https://httr.r-lib.org] son mejores para un control de mayor nivel. El paquete [crul][https://github.com/ropensci/crul/]  es mantenido por rOpenSci. Recomendamos los paquetes mantenidos por rOpenSci [webmockr][https://github.com/ropensci/webmockr/] para imitar las consultas HTTP, y [vcr][https://github.com/ropensci/vcr/] para almacenar en caché las consultas HTTP en los tests del paquete.

Reviewer 2: ```suggestion
- Para las consultas HTTP recomendamos utilizar [curl][https://github.com/jeroen/curl/], [crul][https://github.com/ropensci/crul/], [httr][https://httr.r-lib.org] o [httr2][https://httr2.r-lib.org] en vez de [RCurl][https://cran.rstudio.com/web/packages/RCurl/]. El paquete [curl][https://github.com/jeroen/curl/] es mejor si prefieres los clientes de bajo nivel de HTTP, mientras que [crul][https://github.com/ropensci/crul/] o [httr][https://httr.r-lib.org] son mejores para un control de mayor nivel. El paquete [crul][https://github.com/ropensci/crul/]  es mantenido por rOpenSci. Recomendamos los paquetes mantenidos por rOpenSci [webmockr][https://github.com/ropensci/webmockr/] para imitar las consultas HTTP, y [vcr][https://github.com/ropensci/vcr/] para almacenar en caché las consultas HTTP en los tests del paquete.
```

Reviewer 2: ¿porqué se reemplazaron todos los nombres de los links con las url?  Se debería reemplazar los corchetes [] por paréntesis () para que sea un link funcional?

Reviewer: Alguna cosa no logro entender. En la versión original veo "[curl][]" sin embargo en la web el link funciona correctamente. No se si hay alguna magia de Maelle involucrada. Luego veo que deepl agregó el nombre del paquete en los segundos corchetes y quedó "[curl][curl]" y luego Elio lo reemplazó por los links correspondientes pero capaz se le pasó el tema del los [].

Asi que tendríamos que cambiar los corchetes o entender como es que funciona en la versión en inglés!

Reviewer 2: Ohh, es sintaxis markdown.  Podes nombrar urls.  La forma de escribirlo es [texto para el link][nombre del link] y al final del documento pones

[nombre del link]: url

y cuando knitea te crea los links.

Reviewer: 🤯

Author: Los corchetes son un error! Deberían ser paréntesis para que sean los links normales de markdown.

```suggestion
- Para las consultas HTTP recomendamos utilizar [curl](https://github.com/jeroen/curl/), [crul](https://github.com/ropensci/crul/), [httr](https://httr.r-lib.org) o [httr2](https://httr2.r-lib.org) en vez de [RCurl](https://cran.rstudio.com/web/packages/RCurl/). El paquete [curl](https://github.com/jeroen/curl/) es mejor es mejor si prefieres los clientes de bajo nivel de HTTP, mientras que [crul](https://github.com/ropensci/crul/) o [httr](https://httr.r-lib.org) son mejores para un control de mayor nivel. El paquete [crul](https://github.com/ropensci/crul/) es mantenido por rOpenSci. Recomendamos los paquetes mantenidos por rOpenSci [webmockr](https://github.com/ropensci/webmockr/) para imitar las consultas HTTP, y [vcr](https://github.com/ropensci/vcr/) para almacenar en caché las consultas HTTP en los tests del paquete.
```

- Para parsear JSON, utiliza [jsonlite][https://github.com/jeroen/jsonlite] en lugar de [rjson][rjson] o [RJSONIO][https://cran.rstudio.com/web/packages/RJSONIO/].

- Para parsear, crear y manipular XML, recomendamos fuertemente [xml2][https://cran.rstudio.com/web/packages/xml2/] en la mayoría de los casos. [Puedes consultar las notas de Daniel Nüst sobre la migración de XML a xml2](https://gist.github.com/nuest/3ed3b0057713eb4f4d75d11bb62f2d66).

> - Para datos espaciales, el paquete [sp][https://github.com/edzer/sp/] debe considerarse obsoleto en favor de [sf][https://r-spatial.github.io/sf/]. Los paquetes [rgdal][https://cran.r-project.org/web/packages/rgdal/index.html], [maptools][https://cran.r-project.org/web/packages/maptools/index.html] y [rgeos][https://cran.r-project.org/web/packages/rgeos/index.html] se retirarán a finales de 2023. Recomendamos el uso de las suites espaciales desarrolladas por las comunidades [r-spatial](https://github.com/r-spatial) y [rspatial](https://github.com/rspatial). Consulta [este issue de GitHub](https://github.com/ropensci/software-review-meta/issues/47) para ver las discusiones pertinentes.

Reviewer 2: Si issue es "nena" entonces es [esta issue de GitHub]

Reviewer 2: Queda masculino.

> - Para datos espaciales, el paquete [sp][https://github.com/edzer/sp/] debe considerarse obsoleto en favor de [sf][https://r-spatial.github.io/sf/]. Los paquetes [rgdal][https://cran.r-project.org/web/packages/rgdal/index.html], [maptools][https://cran.r-project.org/web/packages/maptools/index.html] y [rgeos][https://cran.r-project.org/web/packages/rgeos/index.html] se retirarán a finales de 2023. Recomendamos el uso de las suites espaciales desarrolladas por las comunidades [r-spatial](https://github.com/r-spatial) y [rspatial](https://github.com/rspatial). Consulta [este issue de GitHub](https://github.com/ropensci/software-review-meta/issues/47) para ver las discusiones pertinentes.

Reviewer 2: ```suggestion
- Para datos espaciales, el paquete [sp](https://github.com/edzer/sp/) debe considerarse obsoleto en favor de [sf](https://r-spatial.github.io/sf/). Los paquetes [rgdal](https://cran.r-project.org/web/packages/rgdal/index.html), [maptools](https://cran.r-project.org/web/packages/maptools/index.html) y [rgeos](https://cran.r-project.org/web/packages/rgeos/index.html) se retirarán a finales de 2023. Recomendamos el uso de las suites espaciales desarrolladas por las comunidades [r-spatial](https://github.com/r-spatial) y [rspatial](https://github.com/rspatial). Consulta [este issue de GitHub](https://github.com/ropensci/software-review-meta/issues/47) para ver las discusiones pertinentes.
```

Esta es una colección de problemas en CRAN que vale la pena evitar desde el principio.

- Asegúrate de que las palabras del título de tu paquete comiencen con mayúsculas ([*Title Case*](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case)).
- No pongas un punto al final de tu título.
- No pongas "en R" o "con R" en tu título, ya que esto es obvio en los paquetes alojados en CRAN. Si a pesar de todo quieres que esta información aparezca en tu sitio web, revisa la [documentación de `pkgdown`](https://pkgdown.r-lib.org/reference/build_home.html#yaml-config-home) para saber cómo anular esta restricción.
- Evita empezar la descripción con el nombre del paquete o "Este paquete...".
- Asegúrate de incluir enlaces a sitios web si utilizas una API, obtienes datos de una página web, etc... en el campo `Description` de tu archivo `DESCRIPTION`. Las URL deben ir entre simbolos `<>`, por ejemplo `<https://www.r-project.org>`.
- Tanto en el campo `Title` como en `Description` los nombres de los paquetes u otro software externo deben ir entre comillas simples (por ejemplo *Integración de 'Rcpp' para la Biblioteca de Álgebra Lineal Armadillo*).

> - Evita tests y ejemplos que tarden en correr. Considera usar `testthat::skip_on_cran()` en los test que toman mucho tiempo para que se omitan en CRAN pero sigan corriendo localment ey en la [integración continua](#ci).

Reviewer 2: ```suggestion
- Evita tests y ejemplos que tarden en correr. Considera usar `testthat::skip_on_cran()` en los test que toman mucho tiempo para que se omitan en CRAN pero sigan corriendo localmente y en la [integración continua](#ci).
```

- Lee, incorpora y actúa según los consejos del capítulo [*Guía de Colaboración*](#collaboration).

### Aprender sobre el desarrollo de paquetes

#### Libros

- [*R Packages* (Paquetes de R) de Hadley Wickham y Jenny Bryan](https://r-pkgs.org/) es un recurso excelente y fácil de leer sobre el desarrollo de paquetes que está disponible [gratis en línea](https://r-pkgs.org/) (e [impreso - link a la versión anterior de Hadley Wickham, ya que, hasta junio de 2022, la nueva versión aún no se ha publicado](https://www.amazon.com/Packages-Organize-Test-Document-Share/dp/1491910593)).

- [*Writing R Extensions* (Escribiendo extensiones de R)](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) es la referencia canónica, normalmente la más actualizada, para crear paquetes de R.

- [*Mastering Software Development in R* (Dominar el desarrollo de software en R) por Roger D. Peng, Sean Kross y Brooke Anderson](https://bookdown.org/rdpeng/RProgDA/).

- [*Advanced R* (R avanzado) de Hadley Wickham](https://adv-r.hadley.nz/)

#### Tutoriales

> - [La famosa entrada del blog de Hilary Parker *Writing an R package from scratch* (Escribiendo tu propio paquete de R desde cero)](https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/) o [su versión actualizada por Tomas Westlake](https://r-mageddon.netlify.com/post/writing-an-r-package-from-scratch/) que muestra cómo hacer lo mismo de forma más eficiente utilizando `usethis`.

Reviewer 2: ```suggestion
- [El famoso artículo de Hilary Parker *Writing an R package from scratch* (Escribiendo tu propio paquete de R desde cero)](https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/) o [su versión actualizada por Tomas Westlake](https://r-mageddon.netlify.com/post/writing-an-r-package-from-scratch/) que muestra cómo hacer lo mismo de forma más eficiente utilizando `usethis`.
```

- [Blog de R-hub](https://blog.r-hub.io/post).

> - Algunos posts del [blog de rOpenSci](https://ropensci.org/archive/) por ejemplo [*"How to precompute package vignettes or pkgdown articles"* (Cómo pregenerar viñetas delde paquetes o artículo de pkgdown)](https://ropensci.org/blog/2019/12/08/precompute-vignettes/).

Reviewer 2: ```suggestion
- Algunos posts del [blog de rOpenSci](https://ropensci.org/archive/) por ejemplo [*"How to precompute package vignettes or pkgdown articles"* (Cómo pregenerar viñetas de paquetes o artículos de pkgdown)](https://ropensci.org/blog/2019/12/08/precompute-vignettes/).
```
